### PR TITLE
Add comprehensive optimizations to TemplateDslMacro

### PR DESCRIPTION
**Pull request by fomkin**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -139,8 +135,114 @@ val todos = Seq(
 
 renderTodos(todos)
 ```
+# Memory allocation model explanation
 
-Worthy to note
---------------
+As noted below Levsha does not make _additional_
+memory allocation. It possible because every
```
Comment (Tvaroh) [inline README.md]: allocations
It is

File: README.md [hunk 2]
```diff
@@ -139,8 +135,114 @@ val todos = Seq(
...
+template, written in Levsha DSL, in compile time
```
Comment (Tvaroh) [inline README.md]: compile-time

File: README.md [hunk 3]
```diff
@@ -139,8 +135,114 @@ val todos = Seq(
...
+optimizes into calls of `RenderContext` methods
+(unlike a lot another template engines which are
```
Comment (Tvaroh) [inline README.md]: other
which represent

File: README.md [hunk 4]
```diff
@@ -139,8 +135,114 @@ val todos = Seq(
...
+# Memory allocation model explanation
 
-Worthy to note
---------------
+As noted below Levsha does not make _additional_
+memory allocation. It possible because every
+template, written in Levsha DSL, in compile time
+optimizes into calls of `RenderContext` methods
+(unlike a lot another template engines which are 
+represent their templates as AST in heap).
```
Comment (Tvaroh) [inline README.md]: on-heap

File: README.md [hunk 5]
```diff
@@ -139,8 +135,114 @@ val todos = Seq(
...
+  renderContext.addTextNode("Hello world")
+  renderContext.closeNode("h1")
+  renderContext.openNode(XmlNs.html, "p")
+  renderContext.addTextNode("Lorem ipsum dolor")
+  renderContext.closeNode("p")
+  renderContext.closeNode("div")
+}
+```
+
+In turn `RenderContext` (meaning `DiffRenderContext` implementation)
```
Comment (Tvaroh) [inline README.md]: In turn,
meaning -> namely

File: README.md [hunk 6]
```diff
@@ -139,8 +135,114 @@ val todos = Seq(
...
+save instructions in `ByteBuffer` to perform changes inferring in the future.
```
Comment (Tvaroh) [inline README.md]: saves
to infer changes in the future

File: README.md [hunk 7]
```diff
@@ -139,8 +135,114 @@ val todos = Seq(
...
+
+Of course Levsha optimizer does not cover all cases.
```
Comment (Tvaroh) [inline README.md]: Of course,

File: README.md [hunk 8]
```diff
@@ -139,8 +135,114 @@ val todos = Seq(
...
+When optimization can't be performed Levsha just 
+applies current `RenderContext` to unoptimized node.
```
Comment (Tvaroh) [inline README.md]: to the

File: README.md [hunk 9]
```diff
@@ -139,8 +135,114 @@ val todos = Seq(
...
+        }
+    }
+    .foreach { childNode =>
+      childNode.apply(renderContext)
+    }
+  renderContext.closeNode("ul")
+}
+```
+
+When you write your Levsha templates keep in
```
Comment (Tvaroh) [inline README.md]: templates,

File: README.md [hunk 10]
```diff
@@ -139,8 +135,114 @@ val todos = Seq(
...
+  renderContext.closeNode("ul")
+}
+```
+
+When you write your Levsha templates keep in 
+your mind this list of optimizations:
+
+1. Nodes and attrs in branches of `if` expression will be moved to current `RenderContext`
+2. Same for cases of pattern matching
+3. `xs.map(x => 'div(x))` will be rewritten into `while` loop
```
Comment (Tvaroh) [inline README.md]: into a

File: README.md [hunk 11]
```diff
@@ -139,8 +135,114 @@ val todos = Seq(
...
+4. `maybeX.map(x => 'div(x))` will be rewritten into `if` expression
```
Comment (Tvaroh) [inline README.md]: an `if`

File: README.md [hunk 12]
```diff
@@ -139,8 +135,114 @@ val todos = Seq(
...
+5. `void` will be removed
+
+The third point of this list show us how to rewrite
```
Comment (Tvaroh) [inline README.md]: point -> item
shows

File: README.md [hunk 13]
```diff
@@ -139,8 +135,114 @@ val todos = Seq(
...
+previous example so that optimization will perform.
```
Comment (Tvaroh) [inline README.md]: so that optimization could be performed.

File: README.md [hunk 14]
```diff
@@ -139,8 +135,114 @@ val todos = Seq(
...
+    val x = iterator.next()
+    renderContext.openNode(XmlNs.html, "li")
+    renderContext.addTextNode(x.toString)
+    renderContext.closeNode("li")
+  }
+  renderContext.closeNode("div")
+}
+```
+
+If you are not sure about you code run compiler
```
Comment (Tvaroh) [inline README.md]: your code,

**Comment by fomkin** [comment]
@Tvaroh Thanks. Fixed.

**Pull request merged**